lib/types.ts

```diff
@@ -199,9 +199,10 @@ export class TIntersection extends TType {
   public getChecker(suite: ITypeSuite, strict: boolean, allowedProps: Set<string> = new Set()): CheckerFunc {
     const itemCheckers = this.ttypes.map((t) => t.getChecker(suite, strict, allowedProps));
     return (value: any, ctx: IContext) => {
-      const ok = itemCheckers.every(checker => checker(value, ctx))
-      if (ok) { return true; }
-      return ctx.fail(null, null, 0);
+      return itemCheckers.every(checker => {
+        checker(value, ctx.fork());
```

Comment:
Thinking how it's important for `ctx.fork()` and `ctx.completeFork()` to be paired, perhaps a more enforced interface could work nicely, e.g.
```
return ctx.withFork(ctxFork => checker(value, ctxFork))
```

Reply:
This example is a bit exceptional because it returns unconditionally. Usually the code would look like this:

```ts
          const canContinue = ctx.withFork(ctxFork => {
            const ok = propCheckers[i](v, ctxFork);
            if (!ok) { ctxFork.fail(name, null, 1); }
          });
          if (!canContinue) {
            return false;
          }
```

Since you still have to remember to add the conditional return, it doesn't seem to add much benefit.

There's a few things that it would be nice to properly assert to ensure correct usage:

- In `fork()` and in `failed()`, check that the user shouldn't have returned already from a previous failure.
    - For NoopContext, that means `_failed` should be false.
    - For DetailContext, there should be no plain failures (`_propNames` or `_messages`) and not too many `_forks`.
- In `fail()`, check that the user hasn't accidentally called `ctx.fail` where they meant to call `ctxFork.fail` or pass `ctxFork` to a sub checker. This means that since the last call to `fork` (if any) they should have returned and are now in the 'parent' checker where plain failures are allowed again. We can't easily check if they've returned, but they should have: 
    - Called `completeFork()`
    - If `completeFork()` returned `true`, they should have called `failed()`. I forgot to call `failed()` in TIntersection which caused weird results.
- In `completeFork()`, check that there is an active fork, i.e. there was a previous call to `fork()` and no call to `completeFork()` since then.

I don't want these assertions actually living in the code and affecting performance, especially for NoopContext. It seems JS doesn't have assert statements that are automatically removed. I'm thinking of monkey patching these functions in the tests to add the assertions. Is that OK?

Reply:
> Actually the benefit I had in mind wasn't so much to simplify the code that uses it (you are right, it would need as many conditions as before), but rather to make it easier to use correctly.

Yes, but I'm saying the benefit in ensuring correct usage is still minimal at best. I think if you can remember to add `if (...) { return }` then it should be easy from there to fill in the `...` and whether you use `completeFork` or `withFork` makes little difference.

> whenever the usage requires paired calls, it's hard to reason about what happens if an exception happens between them.

Does it matter? If there's an exception then everything is borked and no one will look at the context objects. The word `catch` doesn't appear in this repo. Do you want to catch exceptions in `withFork`? Or is this in case things change significantly in the future?

> I am saying this mainly to clarify my motivation for proposing this alternative interface. And I realize it's internal anyway; and even if it's helpful, it might cost in performance and be undesirable for that reason.

In the end it's not clear to me if:

1. You still want `withFork`
2. You're OK with me monkeypatching `fork/completeFork/withFork/failed` in tests to add assertions.

Reply:
True. I am OK with fork/completeFork/failed. As for monkey-patching assertions, if I understood correctly, this is to verify that they are used correctly by the implementation of other functions, and would only be inside tests, right? I'm fine with that.